Add oauth_body_hash to request headers for non-form-encoded bodies

aparkins opened the pull request:
- Proposal: https://tools.ietf.org/id/draft-eaton-oauth-bodyhash-00.html
 - Certain Oauth1 servers have checks for a hashed body payload header
 that gets included in signing of the request. This is important, as it
 helps avoid a MITM attack-vector where an attacker captures a message,
 replaces the body, and forwards the remaining message with already
 signed headers.
 - Per the linked specification: the oauth_body_hash header is NOT
 provided when a x-www-form-urlencoded body is provided, but SHOULD be
 provided for any other body type.

ribrdb commented:
Should we add a Config option to enable/disable this?
Are we going to need more auther changes to handle the LTI case, since we don't have an access token?

aparkins commented:
Addressing feedback from the previous version of this PR:

> Should we add a Config option to enable/disable this?

Uncertain. The Python OAuth library _doesn't_ have this configuration, but  the `requests_oauthlib` project will strip the body if the `Content-Type` header doesn't match `x-www-form-urlencoded` and the `force_include_body` configuration flag  isn't provided. `pylti` _appears_ to always pass in that flag, except when `ToolOutbound.generate_launch_request()` is called.

It seems like we _probably_ want a layer to perform that same logic for maintaining consistent behavior between the Python and Golang implementations, but I'm not entirely certain what layer would be most appropriate to do that in the Golang code. I'll poke around a bit more, though, and see if I can find something reasonable.

> Are we going to need more auther changes to handle the LTI case, since we don't have an access token?

I don't know? I was under the impression that the only missing aspect between Golang and Python was this request header. Was I misunderstanding?

ribrdb commented:
1) I don't think we need to strictly follow what the python was doing. And we're probably fine having this always on since we're only using this for LTI. But if we want to keep the library general purpose, this extension probably shouldn't be on by default.
2) Well we haven't tried using the library as a client yet because it didn't have request body support. But the library was designed for normal oauth operation, not LTI's signature only stuff so I don't know if that's going to work or not.

aparkins commented:
1. Ah. Well if that's the case, I'm _somewhat_ inclined to enable it by default and disable in the client if we do desire (by manually stripping the body, as `requests_oauthlib` does. It feels appropriate if the general-purpose `oauthlib` has this behavior enabled by default.

2. I can try writing some tests to see what happens. As far as I can tell, the `auther` code doesn't seem bothered by missing those parameters, but I don't exactly know the expected use of these classes to accomplish what we're going to need for LTI.
